modules/sitemaps/sitemap-buffer.php

```diff
@@ -0,0 +1,287 @@
+	 * @param int    $byte_limit The maximum size of the buffer in bytes. Default is 10485760 (10MB).
+	 * @param string $header The string to prepend to the entire buffer.
+	 * @param string $footer The string to append to the entire buffer.
+	 * @param string $time The initial timestamp of the buffer. Must be in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	public function __construct(
+		$item_limit,
+		$byte_limit,
+		$header = '',
+		$footer = '',
+		$time
+	) {
+		$this->item_capacity = max( 1, intval( $item_limit ) );
+
+		mbstring_binary_safe_encoding(); // So we can safely use mb_strlen().
+		$this->byte_capacity = $byte_limit - mb_strlen( $header ) - mb_strlen( $footer );
+		reset_mbstring_encoding();
+
+		$this->footer_text = $footer;
+		$this->buffer = $header;
+		$this->is_full_flag = false;
+		$this->is_empty_flag = true;
+		$this->timestamp = $time;
+		return;
+	}
+
+	/**
+	 * Append an item to the buffer, if there is room for it.
+	 * If not, we set is_full_flag to true. If $item is null,
+	 * don't do anything and report success.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @param string $item The item to be added.
+	 *
+	 * @return bool True if the append succeeded, False if not.
+	 */
+	public function try_to_add_item( $item ) {
+		if ( is_null( $item ) ) {
+			return true;
+		} else {
+			mbstring_binary_safe_encoding(); // So we can safely use mb_strlen().
+
+			if ( 0 >= $this->item_capacity || 0 > $this->byte_capacity - mb_strlen( $item ) ) {
+				$this->is_full_flag = true;
+				reset_mbstring_encoding();
+				return false;
+			} else {
+				$this->is_empty_flag = false;
+				$this->item_capacity -= 1;
+				$this->byte_capacity -= mb_strlen( $item );
+				$this->buffer .= $item;
+				reset_mbstring_encoding();
+				return true;
+			}
+		}
+	}
+
+	/**
+	 * Retrieve the contents of the buffer.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @return string The contents of the buffer (with the footer included).
+	 */
+	public function contents() {
+		return $this->buffer . $this->footer_text;
+	}
+
+	/**
+	 * Detect whether the buffer is full.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @return bool True if the buffer is full, false otherwise.
+	 */
+	public function is_full() {
+		return $this->is_full_flag;
+	}
+
+	/**
+	 * Detect whether the buffer is empty.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @return bool True if the buffer is empty, false otherwise.
+	 */
+	public function is_empty() {
+		return $this->is_empty_flag;
+	}
+
+	/**
+	 * Update the timestamp of the buffer.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @param string $new_time A datetime string in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	public function view_time( $new_time ) {
+		$this->timestamp = max( $this->timestamp, $new_time );
+		return;
+	}
+
+	/**
+	 * Retrieve the timestamp of the buffer.
+	 *
+	 * @since 4.7.0
+	 *
+	 * @return string A datetime string in 'YYYY-MM-DD hh:mm:ss' format.
+	 */
+	public function last_modified() {
+		return $this->timestamp;
+	}
+
+	/**
```

Comment:
Two problems here:

1. `esc_html()` is not the correct escaping function for the html tag name context (the `$open_tag`/`$close_tag` stuff). There actually is no escaping rule for this context; instead we need to sanitize the input so that only allowed characters get through. I don't think core WP has a function for this, but `sanitize_key()` is probably OK.
2. It's probably possible that some non-standard HTML entity (like `&rarr;`) would end up in the output, which is invalid XML. (XML is very strict about what entities are allowed.). See `ent2ncr()`.